extension/source/background/KeyringController.ts

```diff
@@ -101,11 +159,62 @@ export default class KeyringController {
     },
   });
 
-  async BlsWalletAddress(privateKey: string): Promise<string> {
-    return BlsWalletWrapper.Address(
+  async BlsWalletWrapper(privateKey: string): Promise<BlsWalletWrapper> {
+    const blsNetworkConfig = getBlsNetworkConfig(
+      await this.network.read(),
+      this.blsNetworksConfig,
+    );
+
+    return BlsWalletWrapper.connect(
       privateKey,
-      NETWORK_CONFIG.addresses.verificationGateway,
-      this.ethersProvider,
+      blsNetworkConfig.addresses.verificationGateway,
+      await this.ethersProvider.read(),
+    );
+  }
+
+  async getWalletsNetworkData(): Promise<
+    Record<string, { originalGateway: string; address: string } | undefined>
+  > {
+    const walletsNetworkData: Record<
+      string,
+      { originalGateway: string; address: string }
+    > = {};
+
+    const network = await this.network.read();
+
+    const blsNetworkConfig = getBlsNetworkConfig(
+      network,
+      this.blsNetworksConfig,
     );
+
+    const keyring = mixtureCopy(await this.keyring.read());
+    let keyringUpdated = false;
+
+    await Promise.all(
+      keyring.wallets.map(async (wallet) => {
+        let networkDataForWallet = wallet.networks[network.networkKey];
+
+        if (networkDataForWallet === undefined) {
+          networkDataForWallet = {
+            originalGateway: blsNetworkConfig.addresses.verificationGateway,
+            address: await (
+              await this.BlsWalletWrapper(wallet.privateKey)
+            ).address,
+          };
+
+          // eslint-disable-next-line no-param-reassign
+          wallet.networks[network.networkKey] = networkDataForWallet;
```

Comment:
Is there a reason you need to directly assign this here versus using a write/update method on the keyring itself? Seems like there could be some unintended side effects.

Reply:
Yeah sorta. In order to do a write/update you'd need to also do a fresh read beforehand as well. Otherwise when updating another wallet the rest of the keyring will have the old values.

It makes a bit more sense to me this way but I don't feel strongly, happy if you want to commit the update for each wallet.